src/storage/src/hummock/file_cache/store.rs

```diff
@@ -12,431 +12,533 @@
+                        + foyer_store_config.reclaimers / 2,
+                };
+
+                FoyerStore::open(c).await.map_err(FileCacheError::foyer)
+            })
+            .await
+            .unwrap()?;
+
+        Ok(Self::FoyerRuntime {
+            runtime: Arc::new(runtime.into()),
+            store,
+            enable_filter,
+        })
     }
 
-    pub fn cache_file_size(&self) -> usize {
-        self.cache_file.size()
+    #[tracing::instrument(skip(self, value))]
+    pub async fn insert(&self, key: K, value: V) -> Result<bool> {
+        match self {
+            FileCache::None => Ok(false),
+            FileCache::Foyer { store, .. } => store
+                .insert_if_not_exists(key, value)
+                .await
+                .map_err(FileCacheError::foyer),
+            FileCache::FoyerRuntime { runtime, store, .. } => {
+                let store = store.clone();
+                runtime
+                    .spawn(async move { store.insert_if_not_exists(key, value).await })
+                    .await
+                    .unwrap()
+                    .map_err(FileCacheError::foyer)
+            }
+        }
     }
 
-    pub fn cache_file_len(&self) -> usize {
-        self.cache_file.len()
+    #[tracing::instrument(skip(self))]
+    pub fn insert_without_wait(&self, key: K, value: V) {
+        match self {
+            FileCache::None => {}
+            FileCache::Foyer { .. } => panic!("unsupported"),
+            FileCache::FoyerRuntime { runtime, store, .. } => {
+                let store = store.clone();
+                runtime.spawn(async move { store.insert_if_not_exists(key, value).await });
+            }
+        }
     }
 
-    pub fn meta_file_path(&self) -> PathBuf {
-        PathBuf::from(&self.dir).join(META_FILE_FILENAME)
+    /// only fetch value if judge pass
+    #[tracing::instrument(skip(self, fetch_value))]
+    pub async fn insert_with<F, FU>(
+        &self,
+        key: K,
+        fetch_value: F,
+        value_serialized_len: usize,
+    ) -> Result<bool>
+    where
+        F: FnOnce() -> FU,
+        FU: FetchValueFuture<V>,
+    {
+        match self {
+            FileCache::None => Ok(false),
+            FileCache::Foyer { store, .. } => store
+                .insert_if_not_exists_with_future(
+                    key,
+                    fetch_value,
+                    key.serialized_len() + value_serialized_len,
+                )
+                .await
+                .map_err(FileCacheError::foyer),
+            FileCache::FoyerRuntime { runtime, store, .. } => {
+                let store = store.clone();
+                let future = fetch_value();
+                runtime
+                    .spawn(async move {
+                        store
+                            .insert_if_not_exists_with_future(
+                                key,
+                                || future,
```

Comment:
Pass `fetch_value` directly here?

Reply:
It requires `Send + 'static` bound on `F`. Then it will need ownership of `FileCache` to call the method to make sure `&self` outlives `'static`.